### DISCUSSION TITLE
java/datarace/

### DISCUSSION THREAD

**Discussion by giscus** [ambosing/ambosing.github.io#14] [Blog-Comments]
# java/datarace/

아니 이런게 있었다니.. Data Race

https://ambosing.github.io/java/datarace/

**Comment by yunjae228** [comment]
안녕하세요 !! 너무 재밌게 읽었어요 !! 읽다보니 궁금한게 생겼는데
느낌상(?) 위의 상황에서는 한 개의 CPU 코어 내에서 발생하는 일 같은데 맞을까요 ??

그렇다면 여러 개의 코어를 사용하는 상황에서 각 스레드들이 별개의 코어를 사용하게 된다면 어떻게 되나요 ??

너무 재밌었어요 감사합니다 사랑합니다

**Reply by ambosing** [reply]
안녕하세요! 윤재님 ㅎㅎ 
위 상황은 멀티코어든 싱글코어든 공유자원을 가지고 있고, 운영체제나 JVM, 하드웨어에서 비순차적 명령어 처리나 의존성이 없는 코드의 컴파일 순서를 변경하여 최적화 등의 방법을 쓰고 있다면 발생하는 문제입니다! 
그리고 공유자원을 가지고 있어야 합니다!  공유자원을 가지고 있으려면 예를 들어,멀티 프로세스 상황에서는 하나의 파일 안에 변수 값을 두 프로세스가 접근한다거나 멀티 스레드 환경에서는 JAVA에서는 클래스의 필드가 공유자원이므로 이러한 상황에서 발생한다고 볼 수 있습니다 :) ㅎㅎ 
답변이 되었으면 좋겠네요! 요즘 추운데 감기 조심하세요!

**Reply by yunjae228** [reply]
아 !! 이해가 됐습니다 !! 질문을 좀 엉성하게 했군요 ..ㅎㅎ;;;
동기화의 중요성과 어려움을 다시 느끼고 갑니다.. ambosing님도 감기 조심하세요 !
메리 크리스마스 1

**Comment by tldk42** [comment]
파이팅 파이팅!!
